imagecraft/pack/chroot.py

```diff
@@ -0,0 +1,189 @@
+
+        self._mountpoint = base_path / self._relative_mountpoint.lstrip("/")
+        pid = os.getpid()
+        # Only mount if mountpoint exists.
+        emit.debug(f"Mount {self._src} on {str(self._mountpoint)}")
+        logger.debug("[pid=%d] mount %r on chroot", pid, str(self._mountpoint))
+        if self._mountpoint.exists():
+            logger.debug("[pid=%d] mount %r on chroot", pid, str(self._mountpoint))
+            os_utils.mount(self._src, str(self._mountpoint), *args)
+        else:
+            logger.debug(
+                "[pid=%d] mountpoint %r does not exist", pid, str(self._mountpoint)
+            )
+
+    def umount(self, *, lazy: bool = False) -> None:
+        """Umount the mountpoint."""
+        pid = os.getpid()
+
+        if self._mountpoint and self._mountpoint.exists():
+            logger.debug("[pid=%d] umount: %r", pid, str(self._mountpoint))
+
+            os_utils.mount(str(self._mountpoint), "--make-rprivate")
+            args: list[str] = ["--recursive"]
+
+            # Mount points under /dev may be in use and make the bind mount
+            # unmountable. This may happen in destructive mode depending on
+            # the host environment, so use MNT_DETACH to defer unmounting.
+            if lazy:
+                args.append("--lazy")
+            os_utils.umount(str(self._mountpoint), *args)
+
+
+def _runner(
+    path: Path,
+    conn: Connection,
+    target: Callable[..., str | None],
+    args: tuple[str],
+    kwargs: dict[str, Any],
+) -> None:
+    """Chroot to the execution directory and call the target function."""
+    pid = os.getpid()
+    logger.debug("[pid=%d] child process: target=%r", pid, target)
+    try:
+        logger.debug("[pid=%d] chroot to %r", pid, path)
+        os.chdir(path)
+        os.chroot(path)
+        res = target(*args, **kwargs)
+    except Exception as exc:  # noqa: BLE001
+        conn.send((None, str(exc)))
+        return
+
+    conn.send((res, None))
+
+
+class Chroot:
+    """Chroot manager."""
+
+    mounts: list[Mount]
+    path: Path
+
+    def __init__(self, *, path: Path, mounts: list[Mount]) -> None:
+        self.path = path
+        self.mounts = mounts
+
+    def _setup_chroot(self) -> None:
+        """Chroot environment preparation."""
+        # Some images (such as cloudimgs) symlink ``/etc/resolv.conf`` to
+        # ``/run/systemd/resolve/stub-resolv.conf``. We want resolv.conf to be
+        # a regular file to bind-mount the host resolver configuration on.
+        #
+        # There's no need to restore the file to its original condition because
+        # this operation happens on a temporary filesystem layer.
+        logger.debug("setup chroot: %r", self.path)
+        resolv_conf = self.path / "etc/resolv.conf"
+        if resolv_conf.is_symlink():
+            resolv_conf.unlink()
+            resolv_conf.touch()
+        elif not resolv_conf.exists() and resolv_conf.parent.is_dir():
```

Comment:
It looks like we are creating a file if the parent directory exists. What should happen if it does not exist ?

Reply:
This workaround was copied from the original implementation in craft-parts I forked. Since we are now mounting `/run` the name resolution works fine without tweaking this file, so I removed it.